### ISSUE TITLE
Is there any depth fusion code?

### ISSUE THREAD

**Issue by whubaichuan** [kwea123/MVSNet_pl#1]
Good Job.
Is there any depth fusion code?

**Comment by kwea123** [comment]
Hi, currently I don't have time to implement the fusion code. I have provided links of other people's implementation in the readme.

**Comment by whubaichuan** [comment]
@kwea123 Thanks for your reply.
How do you solve the problem about train loss in repo "MVSNet_pytorch"?
I remember you said that the train loss is jumping.

**Comment by kwea123** [comment]
I think it is because the batch size is too small. But eventually it doesn't impact the final performance, so we don't need to worry about it. You can see my training graph that it indeed jumps a lot but finally it converges to good result.

**Comment by whubaichuan** [comment]
@kwea123 ERROR: Package 'pandas' requires a different Python: 3.6.0 not in '>=3.6.1' when I pip install the pytorch-lighting.

Do you have any idear?

**Comment by kwea123** [comment]
Ok it says it requires python>=3.6.1, so can you upgrade your python? I think the latest is 3.6.10 or >=3.7 might work. I recommend using conda environments: install anaconda and do `conda create -n mvsnet_pl python=3.6` then activate it using `conda activate mvsnet_pl`
I will upgrade the readme, thanks.

**Comment by whubaichuan** [comment]
@kwea123 Thanks very much.
I find that your train loss is about 8-9. But the issue "https://github.com/xy-guo/MVSNet_pytorch/issues/6" mentioned that the original MVSNet depth map error is about 4 to 5.

Is there any problem in training? And what does the original MVSNet depth map error  mean?

**Comment by kwea123** [comment]
There are two reasons:
1.  The number of depth sampling: Mine and MVSNet_pytorch's D is 192 and original MVSNet is 256. Larger D makes more precise results.
2.  The fact I use online hard example mining (ohem) https://github.com/kwea123/MVSNet_pl/blob/fc882a59be0d450dea45be11d66cfca8f33dac28/losses.py#L14-L17
    This means I only train on "more difficult regions" where the loss is large, ignoring easy regions. It is a common technique to use in machine learning, so I just add it without any ablation study.

There is no problem in training, if you look at the "absolute error"(=depth map error) instead of loss, it is around 4-5.

**Comment by whubaichuan** [comment]
@kwea123 
Thank your detailed reply.
Would you like to have a try in using your trained model to test the validated dataset?
I see the confusion about the loss in https://github.com/xy-guo/MVSNet_pytorch/issues/6.

**Comment by kwea123** [comment]
I already pasted the val results in [readme](https://github.com/kwea123/MVSNet_pl#metrics). You can also see the validation curves in the log (not pasted here, you can [download](https://github.com/kwea123/MVSNet_pl/releases) it to visualize),  there is no problem, the val loss decreases constantly and the accuracies increase.

**Comment by whubaichuan** [comment]
@kwea123 Thanks for your reply.
The validation curves in the log decrease constantly. That's right. But the curves are in the process of training and the weights of the network are being optimized. Alright?

I mean that using the trained file such ***.ckpt to test the datasets where the weights of the network are loaded by the ***.ckpt file.  Do you know how to solve the confusing loss in https://github.com/xy-guo/MVSNet_pytorch/issues/6.

**Comment by kwea123** [comment]
The numbers in the readme are evaluated using **the final model** where the weights is fixed and not updated anymore. If you evaluate it using the same model you're gonna get the exact same numbers, I just didn't have another piece of code to do only evaluation.
In [xy-guo/MVSNet_pytorch#6](https://github.com/xy-guo/MVSNet_pytorch/issues/6), I think they're looking at the validation curve as well, and his model seems not converging and **I don't know why** so that's why I wrote mine, and my model converges :smile:.

**Comment by whubaichuan** [comment]
@kwea123 Thank for the reply.

I mean that if you log the loss in every batch of every epoch other than the mean loss in every epoch, you will find the loss is abnormal. But the mean loss looks normal at all.

Using loss means that you will get about 30K/2=15K (if you set the batchsize=2) values in one epoch.
Using mean loss means that you will get one value in one epoch.

**Comment by kwea123** [comment]
Please check out my newest repo [CasMVSNet_pl](https://github.com/kwea123/CasMVSNet_pl) which is a improved version of MVSNet. I have written depth fusion code in that repo. If you want to apply it here, it should be very simple with only a minimal change of the depth inference part.

**Comment by whubaichuan** [comment]
@kwea123 hi. Does the depth code  derive from the MVSNet_pytorch？That is a little similar

**Comment by kwea123** [comment]
Yes. I just did a little refactoring and acceleration. It's what's described in MVSNet paper so I didn;t change the concept.

**Comment by whubaichuan** [comment]
@kwea123 hi, I think your code about homo_warp is right. 
And the code about homo_warp is wrong in [here](https://github.com/xy-guo/MVSNet_pytorch/blob/master/models/module.py). He set the init depth_value from 425-900 after rotating the ref image to the camera coordinate of src image. What do you think about the problem?

**Comment by kwea123** [comment]
His implementation is also correct.
I do `R @ [xd yd d].T + T`
He does `(R @ [x y 1].T) * d + T`
They are exactly the same, it's just multiply `d` sooner or later, the result doesn't change.

**Comment by whubaichuan** [comment]
@kwea123 Yes, the depth is scalar.
